src/SharpLearning.Backend.Testing/FlatBatchEnumerator.cs

```diff
@@ -0,0 +1,328 @@
+            TTargetConverter targetConverter)
+        {
+            m_enumerator = enumerator ?? throw new ArgumentNullException(nameof(enumerator));
+            m_targetConverter = targetConverter;
+            m_batchTargets = new TTargetResult[enumerator.TotalBatchSize];
+        }
+
+        public int TotalBatchSize => m_enumerator.TotalBatchSize;
+
+        public bool MoveNext() => m_enumerator.MoveNext();
+
+        public (TFeature[] batchFeatures, TTargetResult[] batchTargets) CurrentBatch()
+        {
+            var (fs, ts) = m_enumerator.CurrentBatch();
+            Transforms.Transform<TTarget, TTargetResult, TTargetConverter>(ts, m_targetConverter, m_batchTargets);
+            return (fs, m_batchTargets);
+        }
+
+        public void Reset() => m_enumerator.Reset();
+    }
+
+    public class OneHotTargetConvertFlatBatchFeaturesTargetEnumerator<TFeature, TTarget, TTargetResult, TTargetIndexer>
+        : IFlatBatchFeaturesTargetEnumerator<TFeature, TTargetResult>
+        where TTargetIndexer : IFunc<TTarget, int>
+    {
+        readonly IFlatBatchFeaturesTargetEnumerator<TFeature, TTarget> m_enumerator;
+        readonly TTargetIndexer m_indexer;
+        readonly TTargetResult m_one;
+        readonly int m_targetCount;
+        // NOTE: This reuses these!
+        readonly TTargetResult[] m_batchTargets;
+
+        public OneHotTargetConvertFlatBatchFeaturesTargetEnumerator(
+            IFlatBatchFeaturesTargetEnumerator<TFeature, TTarget> enumerator,
+            TTargetIndexer indexer,
+            TTargetResult one,
+            int targetCount)
+        {
+            m_enumerator = enumerator ?? throw new ArgumentNullException(nameof(enumerator));
+            m_indexer = indexer;
+            m_one = one;
+            m_targetCount = targetCount;
+            m_batchTargets = new TTargetResult[enumerator.TotalBatchSize * m_targetCount];
+        }
+
+        public int TotalBatchSize => m_enumerator.TotalBatchSize;
+
+        public bool MoveNext() => m_enumerator.MoveNext();
+
+        public (TFeature[] batchFeatures, TTargetResult[] batchTargets) CurrentBatch()
+        {
+            var (fs, ts) = m_enumerator.CurrentBatch();
+            Array.Clear(m_batchTargets, 0, m_batchTargets.Length);
+            for (int i = 0; i < ts.Length; i++)
+            {
+                var index = m_indexer.Invoke(ts[i]);
+                m_batchTargets[i * m_targetCount + index] = m_one;
+            }
+            return (fs, m_batchTargets);
+        }
+
+        public void Reset() => m_enumerator.Reset();
+    }
+
+
+    public class ConvertFlatBatchFeaturesTargetEnumerator<TFeature, TTarget, TFeatureResult, TTargetResult, TFeatureConverter, TTargetConverter>
+        : IFlatBatchFeaturesTargetEnumerator<TFeatureResult, TTargetResult>
+        where TFeatureConverter : IFunc<TFeature, TFeatureResult>
+        where TTargetConverter : IFunc<TTarget, TTargetResult>
```

Comment:
@mdabros just look at this class def, pretty ugly, but for optimal type safety and perf it is needed. Consider when we have to convert GBs of data.

Reply:
I agree, performance should be high priority since this code must be able process and convert many GBs of data efficiently. As long as working with the class from a users perspective is clean, I think it is ok that the class definition is slightly ugly/complicated.